File: android/src/main/java/com/nitrogeolocation/HybridNitroGeolocation.kt
package com.nitrogeolocation

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.Keep
import androidx.core.content.ContextCompat
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitrogeolocation.*

/**
 * Custom exception for permission denied errors
 */
class PermissionDeniedException(message: String) : Exception(message)

@DoNotStrip
@Keep
class HybridNitroGeolocation : HybridNitroGeolocationSpec() {
    
    companion object {
        private const val TAG = "NitroGeolocation"
    }

    private val context: Context
        get() = NitroModules.applicationContext
            ?: throw IllegalStateException("Application context not available")

    // Cached providers - reuse instead of creating new ones each time
    private var fusedProvider: FusedLocationProvider? = null
    private var locationManagerProvider: LocationManagerProvider? = null
    
    // Current active provider for watching
    private var watchProvider: LocationProvider? = null
    
    private var positionListeners: MutableList<(position: GeoPosition) -> Unit> = mutableListOf()
    private var errorListeners: MutableList<(error: GeoError) -> Unit> = mutableListOf()
    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Request location authorization.
     * On Android, this checks the current permission status.
     * Note: Actual permission requests must be done through React Native's PermissionsAndroid API
     * or expo-permissions, as we cannot show permission dialogs from a native module without an Activity.
     */
    override fun requestAuthorization(level: AuthorizationLevel): Promise<AuthorizationResult> {
        return Promise.async {
            // Check if location services are enabled
            val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            val isLocationEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER) ||
                    locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
            
            if (!isLocationEnabled) {
                return@async AuthorizationResult.DISABLED
            }
            
            // Check permission status based on requested level
            val hasFineLocation = ContextCompat.checkSelfPermission(
                context,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
            
            val hasCoarseLocation = ContextCompat.checkSelfPermission(
                context,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
            
            val hasBackgroundLocation = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
                ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.ACCESS_BACKGROUND_LOCATION
                ) == PackageManager.PERMISSION_GRANTED
            } else {
                // Before Android Q, foreground permission implies background access
                hasFineLocation || hasCoarseLocation
            }
            
            when (level) {
                AuthorizationLevel.ALWAYS -> {
                    if ((hasFineLocation || hasCoarseLocation) && hasBackgroundLocation) {
                        AuthorizationResult.GRANTED
                    } else {
                        AuthorizationResult.DENIED
                    }
                }
                AuthorizationLevel.WHENINUSE -> {
                    if (hasFineLocation || hasCoarseLocation) {
                        AuthorizationResult.GRANTED
                    } else {
                        AuthorizationResult.DENIED
                    }
                }
            }
        }
    }

    /**
     * Get or create the appropriate location provider based on options
     * Providers are cached and reused for better performance
     */
    private fun getLocationProvider(forceLocationManager: Boolean): LocationProvider {
        val playServicesAvailable = LocationUtils.isGooglePlayServicesAvailable(context)
        
        return if (forceLocationManager || !playServicesAvailable) {
            locationManagerProvider ?: LocationManagerProvider(context).also {
                Log.i(TAG, "Created LocationManagerProvider")
                locationManagerProvider = it
            }
        } else {
            fusedProvider ?: FusedLocationProvider(context).also {
                Log.i(TAG, "Created FusedLocationProvider")
                fusedProvider = it
            }
        }
    }

    override fun getCurrentPosition(options: GeoOptions): Promise<GeoPosition> {
//        println(fusedProvider)
        return Promise.async {
            if (!LocationUtils.hasLocationPermission(context)) {
                throw PermissionDeniedException("Location permission not granted. Call requestAuthorization() first or request permission via PermissionsAndroid.")
            }

            val forceLocationManager = options.forceLocationManager ?: false
            val provider = getLocationProvider(forceLocationManager)
            
            kotlinx.coroutines.suspendCancellableCoroutine { continuation ->
                provider.getCurrentLocation(
                    options,
                    onSuccess = { location ->
                        if (continuation.isActive) {
                            continuation.resumeWith(
                                Result.success(LocationUtils.locationToGeoPosition(location))
                            )
                        }
                    },
                    onError = { error ->
                        if (continuation.isActive) {
                            continuation.resumeWith(Result.failure(error))
                        }
                    }
                )
                
                // Handle cancellation
                continuation.invokeOnCancellation {
                    Log.d(TAG, "getCurrentPosition cancelled")
                    provider.removeLocationUpdates()
                }
            }
        }
    }

    override fun startObserving(options: GeoWatchOptions) {
        if (!LocationUtils.hasLocationPermission(context)) {
            notifyError(GeoError(PositionError.PERMISSION_DENIED, "Location permission not granted"))
            return
        }

        // Stop any existing observation first
        stopObserving()

        val forceLocationManager = options.forceLocationManager ?: false
        watchProvider = getLocationProvider(forceLocationManager)
        
        watchProvider?.requestLocationUpdates(
            options,
            onLocationChanged = { location ->
                notifyPosition(LocationUtils.locationToGeoPosition(location))
            },
            onError = { error ->
                notifyError(GeoError(PositionError.POSITION_UNAVAILABLE, error.message ?: "Unknown error"))
            }
        )
    }

    override fun stopObserving() {
        watchProvider?.removeLocationUpdates()
        watchProvider = null
    }

    override fun addPositionListener(callback: (position: GeoPosition) -> Unit) {
        positionListeners.add(callback)
    }

    override fun addErrorListener(callback: (error: GeoError) -> Unit) {
        errorListeners.add(callback)
    }

    override fun removeAllListeners() {
        positionListeners.clear()
        errorListeners.clear()
        // Also stop observing when all listeners are removed
        stopObserving()
    }

    private fun notifyPosition(position: GeoPosition) {
        mainHandler.post {
            positionListeners.forEach { listener ->
                try {
                    listener(position)
                } catch (e: Exception) {
                    Log.e(TAG, "Error notifying position listener", e)
                }
            }
        }
    }

    private fun notifyError(error: GeoError) {
        mainHandler.post {
            errorListeners.forEach { listener ->
                try {
                    listener(error)
                } catch (e: Exception) {
                    Log.e(TAG, "Error notifying error listener", e)
                }
            }
        }
    }
    
    /**
     * Clean up all resources - call when the module is being destroyed
     */
    fun cleanup() {
        Log.d(TAG, "Cleaning up NitroGeolocation")
        stopObserving()
        removeAllListeners()
        
        fusedProvider?.removeLocationUpdates()
        fusedProvider = null
        
        locationManagerProvider?.removeLocationUpdates()
        locationManagerProvider = null
    }
}

File: ios/HybridNitroGeolocation.swift
//
//  HybridNitroGeolocation.swift
//  NitroGeolocation
//
//  Created by Rom on 11/26/2025.
//

import Foundation
import CoreLocation
import NitroModules

/**
 * Delegate class to handle authorization status changes
 */
class AuthorizationDelegate: NSObject, CLLocationManagerDelegate {
    private let completion: (CLAuthorizationStatus) -> Void
    
    init(completion: @escaping (CLAuthorizationStatus) -> Void) {
        self.completion = completion
        super.init()
    }
    
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        let status = manager.authorizationStatus
        // Only call completion if status is determined
        if status != .notDetermined {
            completion(status)
        }
    }
}

/**
 * Main HybridObject for NitroGeolocation.
 * Implements the Nitro protocol and delegates to the location provider.
 */
class HybridNitroGeolocation: HybridNitroGeolocationSpec {
    
    private static let TAG = "NitroGeolocation"
    
    // Cached location provider - reused for better performance
    private var locationProvider: CoreLocationProvider?
    
    // Listener callbacks
    private var positionListeners: [(GeoPosition) -> Void] = []
    private var errorListeners: [(GeoError) -> Void] = []
    
    // MARK: - Initialization
    
    override init() {
        super.init()
        log("HybridNitroGeolocation initialized")
    }
    
    deinit {
        cleanup()
        log("HybridNitroGeolocation deinitialized")
    }
    
    // MARK: - Private Methods
    
    /**
     * Get or create the location provider.
     * Provider is cached and reused for better performance.
     */
    private func getLocationProvider() -> CoreLocationProvider {
        if let provider = locationProvider {
            return provider
        }
        
        let provider = CoreLocationProvider()
        locationProvider = provider
        log("Created CoreLocationProvider")
        return provider
    }
    
    private func log(_ message: String) {
        print("[\(HybridNitroGeolocation.TAG)] \(message)")
    }
    
    private func notifyPosition(_ position: GeoPosition) {
        DispatchQueue.main.async { [weak self] in
            self?.positionListeners.forEach { listener in
                listener(position)
            }
        }
    }
    
    private func notifyError(_ error: GeoError) {
        DispatchQueue.main.async { [weak self] in
            self?.errorListeners.forEach { listener in
                listener(error)
            }
        }
    }
    
    /**
     * Clean up all resources.
     */
    private func cleanup() {
        log("Cleaning up NitroGeolocation")
        
        try? stopObserving()
        try? removeAllListeners()
        
        locationProvider?.removeLocationUpdates()
        locationProvider = nil
    }
    
    // MARK: - HybridNitroGeolocationSpec Protocol
    
    func requestAuthorization(level: AuthorizationLevel) throws -> Promise<AuthorizationResult> {
        log("requestAuthorization called with level: \(level)")
        
        let promise = Promise<AuthorizationResult>()
        let locationManager = CLLocationManager()
        
        // Check if location services are enabled
        if !CLLocationManager.locationServicesEnabled() {
            log("Location services disabled")
            promise.resolve(withResult: .disabled)
            return promise
        }
        
        // Check current authorization status
        let currentStatus = locationManager.authorizationStatus
        
        switch currentStatus {
        case .authorizedAlways:
            log("Already authorized always")
            promise.resolve(withResult: .granted)
            
        case .authorizedWhenInUse:
            if level == .wheninuse {
                log("Already authorized when in use")
                promise.resolve(withResult: .granted)
            } else {
                // Need to upgrade to always - request it
                log("Need to upgrade to always authorization")
                requestAuthorizationInternal(level: level, locationManager: locationManager, promise: promise)
            }
            
        case .denied:
            log("Authorization denied")
            promise.resolve(withResult: .denied)
            
        case .restricted:
            log("Authorization restricted")
            promise.resolve(withResult: .restricted)
            
        case .notDetermined:
            log("Authorization not determined, requesting...")
            requestAuthorizationInternal(level: level, locationManager: locationManager, promise: promise)
            
        @unknown default:
            log("Unknown authorization status")
            promise.resolve(withResult: .denied)
        }
        
        return promise
    }
    
    private var authorizationDelegate: AuthorizationDelegate?
    
    private func requestAuthorizationInternal(
        level: AuthorizationLevel,
        locationManager: CLLocationManager,
        promise: Promise<AuthorizationResult>
    ) {
        // Create a delegate to handle the authorization callback
        let delegate = AuthorizationDelegate { [weak self] status in
            self?.log("Authorization callback received: \(status.rawValue)")
            
            switch status {
            case .authorizedAlways:
                promise.resolve(withResult: .granted)
            case .authorizedWhenInUse:
                if level == .wheninuse {
                    promise.resolve(withResult: .granted)
                } else {
                    promise.resolve(withResult: .denied)
                }
            case .denied:
                promise.resolve(withResult: .denied)
            case .restricted:
                promise.resolve(withResult: .restricted)
            case .notDetermined:
                // Shouldn't happen after requesting
                promise.resolve(withResult: .denied)
            @unknown default:
                promise.resolve(withResult: .denied)
            }
            
            // Clean up the delegate
            self?.authorizationDelegate = nil
        }
        
        // Keep a strong reference to the delegate
        authorizationDelegate = delegate
        locationManager.delegate = delegate
        
        // Request the appropriate authorization
        if level == .always {
            locationManager.requestAlwaysAuthorization()
        } else {
            locationManager.requestWhenInUseAuthorization()
        }
    }
    
    func getCurrentPosition(options: GeoOptions) throws -> Promise<GeoPosition> {
        log("getCurrentPosition called")
        
        // Check permissions
        if !LocationUtils.hasLocationPermission() {
            log("Location permission not granted")
            return Promise.rejected(withError: RuntimeError.error(withMessage: "Location permission not granted"))
        }
        
        let provider = getLocationProvider()
        let promise = Promise<GeoPosition>()
        
        provider.getCurrentLocation(
            options: options,
            onSuccess: { location in
                let position = LocationUtils.locationToGeoPosition(location)
                promise.resolve(withResult: position)
            },
            onError: { error in
                promise.reject(withError: RuntimeError.error(withMessage: error.message))
            }
        )
        
        return promise
    }
    
    func startObserving(options: GeoWatchOptions) throws {
        log("startObserving called")
        
        // Check permissions
        if !LocationUtils.hasLocationPermission() {
            log("Location permission not granted")
            notifyError(GeoError(code: .permissionDenied, message: "Location permission not granted"))
            return
        }
        
        // Stop any existing observation first
        try? stopObserving()
        
        let provider = getLocationProvider()
        
        provider.requestLocationUpdates(
            options: options,
            onLocationChanged: { [weak self] location in
                let position = LocationUtils.locationToGeoPosition(location)
                self?.notifyPosition(position)
            },
            onError: { [weak self] error in
                self?.notifyError(error)
            }
        )
    }
    
    func stopObserving() throws {
        log("stopObserving called")
        locationProvider?.removeLocationUpdates()
    }
    
    func addPositionListener(callback: @escaping (GeoPosition) -> Void) throws {
        log("addPositionListener called")
        positionListeners.append(callback)
    }
    
    func addErrorListener(callback: @escaping (GeoError) -> Void) throws {
        log("addErrorListener called")
        errorListeners.append(callback)
    }
    
    func removeAllListeners() throws {
        log("removeAllListeners called")
        positionListeners.removeAll()
        errorListeners.removeAll()
        // Also stop observing when all listeners are removed
        try? stopObserving()
    }
}
